src/main/Client.scala

```diff
@@ -1,35 +1,80 @@
 import org.nlogo.api.ExtensionException
 
-import java.util.concurrent.TimeoutException
+import java.util.concurrent.{ ConcurrentHashMap, Executors, TimeoutException, TimeUnit, Semaphore }
+
+import scala.collection.mutable.{ HashMap => MutableMap },
+  scala.collection.JavaConversions._
 
 import Sockets.MappableSocket
 
 import Messages._
 
 class Client(socketManager: SocketManager) {
-  def rawRequest(address: String, reporter: String): String = {
-    val reqSocket = socketManager.reqSocket(address, {
+  private def requestSocket(address: String) =
+    socketManager.reqSocket(address, {
       s =>
         s.setSendTimeOut(3000)
         s.setReceiveTimeOut(3000)
     })
+
+  private def messageSocket(address: String) =
+    requestSocket(address).mapSend(Request.toString).mapRecv(toResponse)
+
+  private def runRequestReply[A, B](message: B, createdSocket: => MappableSocket[A, B]): A = {
+    val sock = createdSocket
     try {
-      sendRequest(reqSocket, reporter)
-      reqSocket.recv().getOrElse(throw new TimeoutException("response timeout"))
+      sendRequest(sock, message)
+      receiveResponse(sock)
     } finally {
-      reqSocket.close()
+      sock.close()
+    }
+  }
+
+  def request(address: String, req: Request): Response =
+    runRequestReply(req, messageSocket(address))
+
+  def rawRequest(address: String, reqString: String): String =
+    runRequestReply(reqString, requestSocket(address))
+
+  def multiRequest(addresses: Seq[String], req: Request): Seq[Response] = {
```

Comment:
I think this whole function can be reduced to:

```
addresses.par.map(r => request(address, r)).seq
```

That will run all the requests in parallel, erroring if any one of them error.